File: day07/main.cpp
#include <cstdlib>
#include <cmath>
#include <iostream>
#include <vector>
#include <sstream>
#include <fstream>
#include <cstdint>

struct Equation
{
    enum class Operator
    {
        Add = 0,
        Mul,
        NumOperators
    };
    
    enum class OperatorP2
    {
        Add = 0,
        Mul,
        Concat,
        NumOperators
    };
    
    Equation(const std::string& input)
    {
        resultVal = 0;
        std::stringstream ss(input);
        char c;
        ss >> resultVal;
        ss >> c;
        
        uint64_t curVal;
        while (ss >> curVal) {
            values.push_back(curVal);
        }
        if (values.empty()) {
            std::cout << "Failed to parse input values" << std::endl;
            exit(EXIT_FAILURE);
        }
    }
    
    bool IsSolvable()
    {
        if (values.size() == 1) {
            std::cout << "IsSolvable values size == 1" << std::endl;
            exit(EXIT_FAILURE);
        }
        size_t eqNumOperators = values.size()-1;
        size_t numPossibleCombinations = (int)pow((int)Operator::NumOperators,eqNumOperators);
        std::vector<std::vector<Operator>> allCombos(numPossibleCombinations);
        for (auto& vec : allCombos) {
            vec.resize(eqNumOperators);
        }
        size_t numTillSwitch = 1;
        for (int i=eqNumOperators-1; i>=0; i--) {
            int curOp = 0;
            for (size_t j=0; j<numPossibleCombinations; j++) {
                allCombos[j][i] = (Operator)curOp;
                if ((j+1)%numTillSwitch == 0) {
                    curOp = (curOp+1)%(int)Operator::NumOperators;
                }
            }
            numTillSwitch *= (int)Operator::NumOperators;
        }
        for (const std::vector<Operator>& ops : allCombos) {
            if (Evaluate(ops) == resultVal) {
                return true;
            }
        }
        return false;
    }
    
    uint64_t Evaluate(const std::vector<Operator>& operators)
    {
        if (operators.size() != values.size()-1) {
            std::cout << "Unexpected operators/values size" << std::endl;
            exit(EXIT_FAILURE);
        }
        uint64_t curResult = values[0];
        for (size_t i=1; i<values.size(); i++) {
            switch (operators[i-1])
            {
            case Operator::Add:
                curResult = curResult + values[i];
                break;
            case Operator::Mul:
                curResult = curResult * values[i];
                break;
            default:
                std::cout << "Unhandled operator" << std::endl;
                exit(EXIT_FAILURE);
            }
        }
        return curResult;
    }
    
    bool IsSolvableP2()
    {
        if (values.size() == 1) {
            std::cout << "IsSolvable values size == 1" << std::endl;
            exit(EXIT_FAILURE);
        }
        size_t eqNumOperators = values.size()-1;
        size_t numPossibleCombinations = (int)pow((int)OperatorP2::NumOperators,eqNumOperators);
        std::vector<std::vector<OperatorP2>> allCombos(numPossibleCombinations);
        for (auto& vec : allCombos) {
            vec.resize(eqNumOperators);
        }
        size_t numTillSwitch = 1;
        for (int i=eqNumOperators-1; i>=0; i--) {
            int curOp = 0;
            for (size_t j=0; j<numPossibleCombinations; j++) {
                allCombos[j][i] = (OperatorP2)curOp;
                if ((j+1)%numTillSwitch == 0) {
                    curOp = (curOp+1)%(int)OperatorP2::NumOperators;
                }
            }
            numTillSwitch *= (int)OperatorP2::NumOperators;
        }
        for (const std::vector<OperatorP2>& ops : allCombos) {
            if (EvaluateP2(ops) == resultVal) {
                return true;
            }
        }
        return false;
    }
    
    uint64_t EvaluateP2(const std::vector<OperatorP2>& operators)
    {
        if (operators.size() != values.size()-1) {
            std::cout << "Unexpected operators/values size" << std::endl;
            exit(EXIT_FAILURE);
        }
        uint64_t curResult = values[0];
        for (size_t i=1; i<values.size(); i++) {
            switch (operators[i-1])
            {
            case OperatorP2::Add:
                curResult = curResult + values[i];
                break;
            case OperatorP2::Mul:
                curResult = curResult * values[i];
                break;
            case OperatorP2::Concat:
            {
                std::string s1 = std::to_string(curResult);
                std::string s2 = std::to_string(values[i]);
                std::string s3 = s1+s2;
                std::stringstream ss(s3);
                ss >> curResult;
                break;
            }
            default:
                std::cout << "Unhandled operator" << std::endl;
                exit(EXIT_FAILURE);
            }
        }
        return curResult;
    }
    
    uint64_t resultVal;
    std::vector<uint64_t> values;
};
std::vector<Equation> equations;
static void readEquations(const char* fileName)
{
    std::ifstream inFile(fileName);
    if (!inFile || !inFile.is_open()) {
        std::cout << "Failed to open " << fileName << std::endl;
        exit(EXIT_FAILURE);
    }
    std::string s;
    while (std::getline(inFile, s)) {
        equations.push_back(Equation(s));
        s = "";
    }
}

static void part1()
{
    uint64_t valuesSum = 0;
    for (Equation& eq : equations) {
        if (eq.IsSolvable()) {
            valuesSum += eq.resultVal;
        }
    }
    std::cout << "Part1 sum: " << valuesSum << std::endl;
}

static void part2()
{
    uint64_t valueSum = 0;
    for (Equation& eq : equations) {
        if (eq.IsSolvableP2()) {
            valueSum += eq.resultVal;
        }
    }
    std::cout << "Part2 sum: " << valueSum << std::endl;
}

int main(int argc, char* argv[])
{
    if (argc != 2) {
        std::cout << "Usage: " << argv[0] << " <input file>" << std::endl;
        return 1;
    }
    readEquations(argv[1]);
    
    part1();
    part2();
}

File: day07/src/main.rs
use std::env;
use std::fs;

#[derive(Clone,PartialEq,Copy,Debug)]
enum Operator {
    Add,
    Mul
}
const NUM_OPS: u32 = 2;

#[derive(Clone,PartialEq,Copy,Debug)]
enum OperatorP2 {
    Add,
    Mul,
    Concat
}
const NUM_OPS_P2: u32 = 3;

#[derive(Debug)]
struct Equation {
    result_val: u64,
    values: Vec<u64>,
}
impl Equation {
    fn is_solvable(&self) -> bool {
        if self.values.len() == 1 {
            println!("is_solvable values size == 1");
            return false;
        }
        
        let eq_num_ops = self.values.len()-1;
        let num_combos: u32 = NUM_OPS.pow(eq_num_ops.try_into().unwrap());

        let mut all_combos: Vec<Vec<Operator>> = vec![vec![]; num_combos.try_into().unwrap()];
        for i in 0..all_combos.len() {
            all_combos[i].resize(eq_num_ops, Operator::Add);
        }
        let mut num_till_switch: u32 = 1;
        for i in (0..eq_num_ops).rev() {
            let mut cur_op: Operator = Operator::Add;
            for j in 0..num_combos {
                all_combos[j as usize][i] = cur_op;
                if (j+1)%num_till_switch == 0 {
                    match cur_op {
                        Operator::Add => cur_op = Operator::Mul,
                        Operator::Mul => cur_op = Operator::Add
                    }
                }
            }
            num_till_switch *= NUM_OPS;
        }

        for ops in all_combos {
            if self.evaluate(&ops) == self.result_val {
                return true;
            }
        }

        return false;
    }
    fn evaluate(&self, ops: &Vec<Operator>) -> u64 {
        if ops.len() != self.values.len()-1 {
            println!("Unexpected ops len");
            return 0;
        }
        let mut cur_result: u64 = self.values[0];
        for i in 1..self.values.len() {
            match ops[i-1] {
                Operator::Add => cur_result = cur_result + self.values[i],
                Operator::Mul => cur_result = cur_result * self.values[i],
            }
        }

        cur_result
    }

    fn is_solvable_p2(&self) -> bool {
        if self.values.len() == 1 {
            println!("is_solvable values size == 1");
            return false;
        }
        
        let eq_num_ops = self.values.len()-1;
        let num_combos: u32 = NUM_OPS_P2.pow(eq_num_ops.try_into().unwrap());

        let mut all_combos: Vec<Vec<OperatorP2>> = vec![vec![]; num_combos.try_into().unwrap()];
        for i in 0..all_combos.len() {
            all_combos[i].resize(eq_num_ops, OperatorP2::Add);
        }
        let mut num_till_switch: u32 = 1;
        for i in (0..eq_num_ops).rev() {
            let mut cur_op: OperatorP2 = OperatorP2::Add;
            for j in 0..num_combos {
                all_combos[j as usize][i] = cur_op;
                if (j+1)%num_till_switch == 0 {
                    match cur_op {
                        OperatorP2::Add => cur_op = OperatorP2::Mul,
                        OperatorP2::Mul => cur_op = OperatorP2::Concat,
                        OperatorP2::Concat => cur_op = OperatorP2::Add
                    }
                }
            }
            num_till_switch *= NUM_OPS_P2;
        }

        for ops in all_combos {
            if self.evaluate_p2(&ops) == self.result_val {
                return true;
            }
        }

        return false;
    }
    fn evaluate_p2(&self, ops: &Vec<OperatorP2>) -> u64 {
        if ops.len() != self.values.len()-1 {
            println!("Unexpected ops len");
            return 0;
        }
        let mut cur_result: u64 = self.values[0];
        for i in 1..self.values.len() {
            match ops[i-1] {
                OperatorP2::Add => cur_result = cur_result + self.values[i],
                OperatorP2::Mul => cur_result = cur_result * self.values[i],
                OperatorP2::Concat => {
                    let str1: String = cur_result.to_string();
                    let str2: String = self.values[i].to_string();
                    let str3: String = str1 + &str2;
                    let new_val: u64 = str3.parse().unwrap();
                    cur_result = new_val;
                }
            }
        }

        cur_result
    }
}

fn build_equation(in_str: &String) -> Equation {
    let mut tokens = in_str.split(':');
    let result_val = sscanf::sscanf!(tokens.next().unwrap(), "{}", u64).unwrap();

    let values: Vec<u64> = tokens.next().unwrap().split(' ')
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .map(|s| s.parse().unwrap())
        .collect();

    Equation {
        result_val,
        values
    }
}

fn read_input_eqs(file_name: &String) -> Vec<Equation> {
    return fs::read_to_string(file_name)
        .unwrap()
        .lines()
        .map(String::from)
        .map(|s| build_equation(&s))
        .collect();
}

fn part1(eqs: &Vec<Equation>) {
    let mut vals_sum: u64 = 0;
    for eq in eqs {
        if eq.is_solvable() {
            vals_sum += eq.result_val;
        }
    }
    println!("Part1 sum: {}", vals_sum);
}

fn part2(eqs: &Vec<Equation>) {
    let mut vals_sum: u64 = 0;
    for eq in eqs {
        if eq.is_solvable_p2() {
            vals_sum += eq.result_val;
        }
    }
    println!("Part2 sum: {}", vals_sum);
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: {} <filename>", args[0]);
        return;
    }
    let eqs = read_input_eqs(&args[1]);
    
    part1(&eqs);
    part2(&eqs);
}

